compiler/lcalc/compile_without_exceptions.ml

```diff
@@ -0,0 +1,455 @@
+      in
+      (e', cuts)
+  | EArray es ->
+      let es', cuts = es |> List.map (translate_and_cut ctx) |> List.split in
+
+      ( Bindlib.box_apply (fun es' -> (A.EArray es', pos)) (Bindlib.box_list es'),
+        disjoint_union_maps pos cuts )
+  | EOp op -> (Bindlib.box (A.EOp op, pos), A.VarMap.empty)
+
+and translate_expr ?(append_esome = true) (ctx : ctx) (e : D.expr Pos.marked) :
+    A.expr Pos.marked Bindlib.box =
+  let e', cs = translate_and_cut ctx e in
+  let cs = A.VarMap.bindings cs in
+
+  let _pos = Pos.get_position e in
+
+  (* build the cuts *)
+  Cli.debug_print
+  @@ Format.asprintf "cut for the expression: [%a]" (Format.pp_print_list pp_var) (List.map fst cs);
+
+  ListLabels.fold_left cs
+    ~init:(if append_esome then A.make_some e' else e')
+    ~f:(fun acc (v, (c, pos_c)) ->
+      Cli.debug_print @@ Format.asprintf "cut using A.%a" pp_var v;
+
+      let c' : A.expr Pos.marked Bindlib.box =
+        match c with
+        (* Here we have to handle only the cases appearing in cuts, as defined the [translate_and_cut] function. *)
+        | D.EVar v -> (find ~info:"should never happend" (Pos.unmark v) ctx).expr
+        | D.EDefault (excep, just, cons) ->
+            let excep' = List.map (translate_expr ctx) excep in
+            let just' = translate_expr ctx just in
+            let cons' = translate_expr ctx cons in
+            (* calls handle_option. *)
+            A.make_app
+              (A.make_var (A.handle_default_opt, pos_c))
+              [
+                Bindlib.box_apply (fun excep' -> (A.EArray excep', pos_c)) (Bindlib.box_list excep');
+                just';
+                cons';
+              ]
+              pos_c
+        | D.ELit D.LEmptyError -> A.make_none pos_c
+        | D.EAssert arg ->
+            let arg' = translate_expr ctx arg in
+
+            (* [
+                 match arg with
+                 | None -> raise NoValueProvided
+                 | Some v -> assert {{ v }}
+               ] *)
+            let silent_var = A.Var.make ("_", pos_c) in
+            let x = A.Var.make ("assertion_argument", pos_c) in
+
+            A.make_matchopt_dumb arg'
+              (A.make_abs [| silent_var |]
+                 (Bindlib.box (A.ERaise A.NoValueProvided, pos_c))
+                 pos_c [ (D.TAny, pos_c) ] pos_c)
+              (A.make_abs [| x |]
+                 (Bindlib.box_apply (fun arg -> (A.EAssert arg, pos_c)) (A.make_var (x, pos_c)))
+                 pos_c [ (D.TAny, pos_c) ] pos_c)
+        | _ ->
+            Errors.raise_spanned_error
+              "Internal Error: An term was found in a position where it should not be" pos_c
+      in
+
+      (* [
+           match {{ c' }} with
+           | None -> None
+           | Some {{ v }} -> {{ acc }}
+           end
+         ] *)
+      Cli.debug_print @@ Format.asprintf "build matchopt using %a" pp_var v;
+      A.make_matchopt pos_c v (D.TAny, pos_c) c' (A.make_none pos_c) acc)
+
+type scope_lets =
```

Comment:
If this works, then we should make it the default in `Dcalc.Ast`.

Reply:
This is noted, but will require some work on the z3/verification side.

Reply:
Ok, can you file an issue with details about what to rewrite later?

Reply:
#190